storage: invalidate cached lastTerm on snapshots

nvb opened the pull request:
Addresses the current issue in #17524.

I'll open an issue to properly test this, but that might take some time.
Fow now this seems like a clear fix that we should get in sooner rather
than later.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18327)
<!-- Reviewable:end -->

tschottdorf commented on pkg/storage/replica_raftstorage.go:
```diff
@@ -814,15 +821,7 @@ func (r *Replica) applySnapshot(
-	// raftpb.SnapshotMetadata.
-	r.mu.lastIndex = s.RaftAppliedIndex
-	r.mu.lastTerm = 0
+	r.mu.lastIndex, r.mu.lastTerm = lastIndexAndTermInSnapshot(inSnap)
```
couldn't you just return `struct lastIndexAndTerm` from `applySnapshot` and save the other call to `lastIndexAndTermInSnapshot`? Nit, but I think it might be clearer.

tbg reviewed (approved):
Looks good, but it strikes me as brittle that we have two values that are tied together but are kept separate.

Maybe a `struct lastTermAndIndex` could make it more obvious that changing one without the other is illegal. Just a thought though.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo not changing where we retrieve `lastIndex` from following a snapshot.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 3254 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18327#-KtSAXXg5wm--bUmd8E5:-KtSAXXg5wm--bUmd8E6:b-qcpcch) ([raw file](https://github.com/cockroachdb/cockroach/blob/91a299dd6205c23cb02fa601c3500850cb93605a/pkg/storage/replica.go#L3254)):*
> ```Go
> 		if lastIndex, err = r.raftMu.stateLoader.loadLastIndex(ctx, r.store.Engine()); err != nil {
> 			return stats, err
> 		}
> ```

For a 1.1 cherry-pick, I'd prefer doing `lastTerm = invalidLastTerm` and stick with loading `lastIndex` from disk. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18327#-:-KtSAmxUpxzFzMdcZNZi:b-xzunzc)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Heh, two opposite recommendations. Should I split it up into two commits, one with just the one line change and one with the rest? Then we can just cherry-pick in the first one.

tschottdorf commented:
I don't feel strongly, we could also just do the beauty refactor later in
the cycle if we track it. Getting a nice contained cherry-pick is the more
important item.

On Thu, Sep 7, 2017 at 12:40 PM Nathan VanBenschoten <
notifications@github.com> wrote:

> Heh, two opposite recommendations. Should I split it up into two commits,
> one with just the one line change and one with the rest? Then we can just
> cherry-pick in the first one.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/18327#issuecomment-327855809>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135CcDLYCvbyWfkFySvei4iL12sZn7ks5sgBxwgaJpZM4PQD4f>
> .
>
-- 

-- Tobias

petermattis commented:
> I don't feel strongly, we could also just do the beauty refactor later in
the cycle if we track it. Getting a nice contained cherry-pick is the more
important item.

Agreed on the nicely contained cherry-pick being the priority. That's what @nvanbenschoten and i discussed in person. I think it should be followed up quickly with both the beauty refactor and a test.

nvb committed:
storage: invalidate cached lastTerm on snapshots

Addresses the current issue in #17524.

I'll open an issue to properly test this, but that might take some time.
Fow now this seems like a clear fix that we should get in sooner rather
than later.

nvanbenschoten commented:
Done. PTAL.

Opened #18336 to track cleanup and testing.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18327#-:-KtShmvnLGFi19Lerm7t:bnfp4nl)*
<!-- Sent from Reviewable.io -->
